File: app/src/main/java/ca/rmen/android/poetassistant/main/dictionaries/rt/Rhymer.java
/*
 * Copyright (c) 2016 Carmen Alvarez
 *
 * This file is part of Poet Assistant.
 *
 * Poet Assistant is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Poet Assistant is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Poet Assistant.  If not, see <http://www.gnu.org/licenses/>.
 */

package ca.rmen.android.poetassistant.main.dictionaries.rt;

import android.database.Cursor;
import android.support.annotation.NonNull;
import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.SortedSet;
import java.util.TreeSet;

import javax.inject.Inject;

import ca.rmen.android.poetassistant.Constants;
import ca.rmen.android.poetassistant.main.dictionaries.EmbeddedDb;
import ca.rmen.android.poetassistant.settings.SettingsPrefs;
import ca.rmen.rhymer.RhymeResult;
import ca.rmen.rhymer.WordVariant;

public class Rhymer extends ca.rmen.rhymer.Rhymer {
    private static final String TAG = Constants.TAG + Rhymer.class.getSimpleName();

    private final EmbeddedDb mEmbeddedDb;
    private final SettingsPrefs mPrefs;

    @Inject
    public Rhymer(EmbeddedDb embeddedDb, SettingsPrefs prefs) {
        mEmbeddedDb = embeddedDb;
        mPrefs = prefs;
    }

    public boolean isLoaded() {
        return mEmbeddedDb.isLoaded();
    }

    @NonNull
    @Override
    protected List<WordVariant> getWordVariants(String word) {
        List<WordVariant> result = new ArrayList<>();
        String[] projection = new String[]{"variant_number", "stress_syllables", "last_syllable", "last_two_syllables", "last_three_syllables"};
        String selection = "word=?";
        String[] selectionArgs = new String[]{word};
        Cursor cursor = mEmbeddedDb.query("word_variants", projection, selection, selectionArgs);
        if (cursor != null) {
            try {
                while (cursor.moveToNext()) {
                    int column = 0;
                    int variantNumber = cursor.getInt(column++);
                    String lastStressSyllable = cursor.getString(column++);
                    String lastSyllable = cursor.getString(column++);
                    String lastTwoSyllables = cursor.getString(column++);
                    String lastThreeSyllables = cursor.getString(column);
                    WordVariant wordVariant = new WordVariant(variantNumber, lastStressSyllable, lastSyllable, lastTwoSyllables, lastThreeSyllables);
                    result.add(wordVariant);
                }
            } finally {
                cursor.close();
            }
        }
        return result;
    }


    /**
     * @return the words which rhyme with the given word, in any order, matching one, two or three
     * syllables.
     */
    Set<String> getFlatRhymes(String word) {
        List<RhymeResult> rhymeResults = super.getRhymingWords(word);
        Set<String> flatRhymes = new HashSet<>();
        for (RhymeResult rhymeResult : rhymeResults) {
            Collections.addAll(flatRhymes, rhymeResult.strictRhymes);
            Collections.addAll(flatRhymes, rhymeResult.oneSyllableRhymes);
            Collections.addAll(flatRhymes, rhymeResult.twoSyllableRhymes);
            Collections.addAll(flatRhymes, rhymeResult.threeSyllableRhymes);
        }
        return flatRhymes;
    }

    @Override
    protected SortedSet<String> getWordsWithLastStressSyllable(String syllable) {
        return lookupBySyllable(syllable, "stress_syllables");
    }

    @Override
    protected SortedSet<String> getWordsWithLastSyllable(String syllable) {
        return lookupBySyllable(syllable, "last_syllable");
    }

    @Override
    protected SortedSet<String> getWordsWithLastTwoSyllables(String syllables) {
        return lookupBySyllable(syllables, "last_two_syllables");
    }

    @Override
    protected SortedSet<String> getWordsWithLastThreeSyllables(String syllables) {
        return lookupBySyllable(syllables, "last_three_syllables");
    }

    /**
     * Of the given words, returns a set containing those which have a definition in the dictionary table.
     */
    Set<String> getWordsWithDefinitions(String[] words) {
        if (words.length == 0) return Collections.emptySet();
        Log.v(TAG, "getWordsWithDefinitions for " + words.length + " words");
        Set<String> result = new HashSet<>();
        String[] projection = new String[]{"word"};
        int queryCount = EmbeddedDb.getQueryCount(words.length);
        for (int i = 0; i < queryCount; i++) {
            String[] queryWords = EmbeddedDb.getArgsInQuery(words, i);
            Log.v(TAG, "getWordsWithDefinitions: query " + i + " has " + queryWords.length + " words");
            String selection = "word in " + buildInClause(queryWords.length) + " AND has_definition=1";
            Cursor cursor = mEmbeddedDb.query("word_variants", projection, selection, queryWords);
            if (cursor != null) {
                try {
                    while (cursor.moveToNext()) {
                        String word = cursor.getString(0);
                        result.add(word);
                    }
                } finally {
                    cursor.close();
                }
            }
        }

        return result;
    }


    private static String buildInClause(int size) {
        StringBuilder builder = new StringBuilder(size * 2 + 1);
        builder.append('(');
        for (int i = 0; i < size; i++) {
            builder.append('?');
            if (i != size - 1) builder.append(',');
        }
        builder.append(')');
        return builder.toString();
    }

    @NonNull
    private SortedSet<String> lookupBySyllable(String syllables, String columnName) {
        SortedSet<String> result = new TreeSet<>();
        String[] projection = new String[]{"word"};
        String selection = columnName + "=?";
        if (!mPrefs.getIsAllRhymesEnabled()) {
            selection += "AND has_definition=1";
        }
        String[] selectionArgs = new String[]{syllables};
        Cursor cursor = mEmbeddedDb.query("word_variants", projection, selection, selectionArgs);
        if (cursor != null) {
            try {
                while (cursor.moveToNext()) {
                    String word = cursor.getString(0);
                    result.add(word);
                }
            } finally {
                cursor.close();
            }
        }
        return result;

    }
}

File: app/src/main/kotlin/ca/rmen/android/poetassistant/main/dictionaries/rt/Rhymer.kt
/*
 * Copyright (c) 2016-2017 Carmen Alvarez
 *
 * This file is part of Poet Assistant.
 *
 * Poet Assistant is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Poet Assistant is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Poet Assistant.  If not, see <http://www.gnu.org/licenses/>.
 */

package ca.rmen.android.poetassistant.main.dictionaries.rt

import android.util.Log
import ca.rmen.android.poetassistant.Constants
import ca.rmen.android.poetassistant.main.dictionaries.EmbeddedDb
import ca.rmen.android.poetassistant.settings.SettingsPrefs
import ca.rmen.rhymer.WordVariant
import java.util.Locale
import java.util.SortedSet
import java.util.TreeSet
import javax.inject.Inject

class Rhymer @Inject constructor(private val embeddedDb: EmbeddedDb, private val prefs: SettingsPrefs) : ca.rmen.rhymer.Rhymer() {
    companion object {
        private val TAG = Constants.TAG + Rhymer::class.java.simpleName
    }

    fun isLoaded(): Boolean = embeddedDb.isLoaded()

    override fun getWordVariants(word: String): List<WordVariant> {
        val result = ArrayList<WordVariant>()
        val projection = arrayOf("variant_number", "stress_syllables", "last_syllable", "last_two_syllables", "last_three_syllables")
        val selection = "word=?"
        val selectionArgs = arrayOf(word)
        embeddedDb.query("word_variants", projection, selection, selectionArgs)?.use { cursor ->
            while (cursor.moveToNext()) {
                var column = 0
                val variantNumber = cursor.getInt(column++)
                val lastStressSyllable = cursor.getString(column++)
                val lastSyllable = cursor.getString(column++)
                val lastTwoSyllables = cursor.getString(column++)
                val lastThreeSyllables = cursor.getString(column)
                val wordVariant = WordVariant(variantNumber, lastStressSyllable, lastSyllable, lastTwoSyllables, lastThreeSyllables)
                result.add(wordVariant)
            }
        }
        return result
    }

    /**
     * @return the words which rhyme with the given word, in any order, matching one, two or three
     * syllables.
     */
    fun getFlatRhymes(word: String): Set<String> {
        val rhymeResults = super.getRhymingWords(word, Constants.MAX_RESULTS)
        val flatRhymes = HashSet<String>()
        rhymeResults.forEach {
            flatRhymes.addAll(it.strictRhymes)
            flatRhymes.addAll(it.oneSyllableRhymes)
            flatRhymes.addAll(it.twoSyllableRhymes)
            flatRhymes.addAll(it.threeSyllableRhymes)
        }
        return flatRhymes
    }

    override fun getWordsWithLastStressSyllable(lastStressSyllable: String): SortedSet<String> {
        return lookupBySyllable(lastStressSyllable, "stress_syllables")
    }

    override fun getWordsWithLastSyllable(lastSyllable: String): SortedSet<String> {
        return lookupBySyllable(lastSyllable, "last_syllable")
    }

    override fun getWordsWithLastTwoSyllables(lastTwoSyllables: String): SortedSet<String> {
        return lookupBySyllable(lastTwoSyllables, "last_two_syllables")
    }

    override fun getWordsWithLastThreeSyllables(lastThreeSyllables: String): SortedSet<String> {
        return lookupBySyllable(lastThreeSyllables, "last_three_syllables")
    }

    /**
     * Of the given words, returns a set containing those which have a definition in the dictionary table.
     */
    fun getWordsWithDefinitions(words: Array<String>): Set<String> {
        if (words.isEmpty()) return emptySet()
        Log.v(TAG, "getWordsWithDefinitions for ${words.size} words")
        val result = HashSet<String>()
        val projection = arrayOf("word")
        val queryCount = EmbeddedDb.getQueryCount(words.size)
        for (i in 0 until queryCount) {
            val queryWords = EmbeddedDb.getArgsInQuery(words, i)
            Log.v(TAG, "getWordsWithDefinitions: query $i has ${queryWords.size} words")
            val selection = "word in " + EmbeddedDb.buildInClause(queryWords.size) + " AND has_definition=1"
            embeddedDb.query("word_variants", projection, selection, queryWords)?.use { cursor ->
                while (cursor.moveToNext()) {
                    result.add(cursor.getString(0))
                }
            }
        }
        return result
    }

    private fun lookupBySyllable(syllables: String, columnName: String): SortedSet<String> {
        val result = TreeSet<String>()
        val projection = arrayOf("word")
        var selectionColumn = columnName
        var inputSyllables = syllables
        if (prefs.isAORAOMatchEnabled && syllables.contains("AO")) {
            selectionColumn = String.format(Locale.US, "replace(%s, 'AOR', 'AO'", selectionColumn)
            inputSyllables = inputSyllables.replace("AOR", "AO")
        }
        if (prefs.isAOAAMatchEnabled && (syllables.contains("AO") || syllables.contains("AA"))) {
            selectionColumn = String.format(Locale.US, "replace(%s, 'AO', 'AA')", selectionColumn)
            inputSyllables = inputSyllables.replace("AO", "AA")
        }
        var selection = selectionColumn + " = ? "
        if (!prefs.isAllRhymesEnabled) {
            selection += "AND has_definition=1"
        }
        val selectionArgs = arrayOf(inputSyllables)
        embeddedDb.query("word_variants", projection, selection, selectionArgs)?.use { cursor ->
            while (cursor.moveToNext()) {
                result.add(cursor.getString(0))
            }
        }
        return result
    }
}
